### ISSUE TITLE
Field Hooks are not applied to embedded struct fields

### ISSUE THREAD

**Issue by willhug** [uber-go/mapdecode#15]
Field Hooks allow interpolation on struct fields (most commonly used to add environment variables to strings).  So if I have a struct:

```
type Config struct {
   MyField: `config:"field,interpolate"`
}
```

I can specify the config field as `config: hello${MYENVVAR:backup}` which will use the environment variable if it is available or fallback to "backup".

The issue happens if I embed structs.  So if I have:

```
type Config struct {
  SubConfig
}
type SubConfig struct {
  MyField: `config:"field,interpolate"`
}
```

The code for Field Hooks does not properly interpolate the `MyField` struct field (i.e. we don't interpolate env vars).

I've narrowed this down to this for loop when determining to interpolate fields:
https://github.com/uber-go/mapdecode/blob/master/hooks.go#L233

This for loop does not account for embedded structs.  

We can compare this to the `decodeStruct` method that iterates over embedded structs here:
https://github.com/uber-go/mapdecode/blob/master/internal/mapstructure/mapstructure.go#L701

Proposal: We should update the FieldHook function to iterate over the embedded structs similar to the `decodeStruct` method

Thoughts: @prashantv @abhinav @breerly @akshayjshah 
(Mostly wondering if this was a conscious choice before implementing)

**Comment by abhinav** [comment]
This is definitely unintentional. We can update that loop in hooks.go or we can implement field hooks natively in `decodeStruct` so that we don't have to keep the field iteration logic in sync. I suggest going for the latter.

**Comment by abhinav** [comment]
Resolved by #19
